R/functions-Spectrum.R

```diff
@@ -915,7 +935,7 @@ descendPeak <- function(mz, intensity, peakIdx = NULL, signalPercentage = 33,
 #'     col = "black")
 combineSpectra <- function(x, mzFun = base::mean, intensityFun = base::mean,
                            main = floor(length(x) / 2L) + 1L, mzd,
-                           timeDomain = TRUE) {
+                           timeDomain = FALSE, onlyMain = FALSE) {
```

Comment:
Good point! Thanks @sgibb! I opt for `unionPeaks`. Regarding the `main`, we need that one, because I have to define a `Spectrum` from which I want to keep all the other information (such as retention time etc.

Reply:
> Regarding the main, we need that one, because I have to define a Spectrum from which I want to keep all the other information (such as retention time etc.

I understand but that doesn't have to be user visible, I mean something like:
```r
combineSpectra <- function(x, main=floor(length(x) / 2L) + 1L, ...) {
    if (is.null(main))
        main <- 1 # or floor(length(x) / 2L) + 1L
    # ...
}
```

It is just to avoid growing number of arguments. But I am fine with `unionPeaks`.

Reply:
If I understand you correctly you would like to replace `main = floor(...` in the function definitions with `main = NULL`? I have no strong opinion there - but this would add an additional `if` statement within the function.

Reply:
No, the default would be still `main=floor(...` and the `unionPeaks` argument will be removed. And IMHO there is no need to use an additional `if` statement:
```r
combineSpectra <- function(x, mzFun = base::mean, intensityFun = base::mean,
                           main = floor(length(x) / 2L) + 1L, mzd,
                           timeDomain = FALSE) {

    if (length(unique(unlist(lapply(x, function(z) z@msLevel)))) != 1)
        stop("Can only combine spectra with the same MS level")
    mzs <- lapply(x, function(z) z@mz)
    mzs_lens <- base::lengths(mzs)
    mzs <- unlist(mzs, use.names = FALSE)
    mz_order <- base::order(mzs)
    mzs <- mzs[mz_order]
    if (timeDomain)
        mz_groups <- .group_mz_values(sqrt(mzs), mzd = mzd)
    else
mz_groups <- .group_mz_values(mzs, mzd = mzd)
    
    if (is.null(main) || length(x) > main || main < 1) {
        main <- floor(length(x) / 2L) + 1L
    } else {
        ## Want to keep only those groups with a m/z from the main spectrum.
        ## vectorized version from @sgibb
        is_in_main <- rep.int(seq_along(mzs_lens), mzs_lens)[mz_order] == main
        keep <- mz_groups %in% mz_groups[is_in_main]
        ## Keep only values for which a m/z in main is present.
        mz_groups <- mz_groups[keep]
        mzs <- mzs[keep]
        ints <- ints[keep]
    }
    if (length(unique(mz_groups)) < length(x[[main]]@mz))
        warning("Got less m/z groups than m/z values in the original spectrum.",
                " Most likely the data is not profile-mode LCMS data.")
    ints <- unlist(base::lapply(x, function(z) z@intensity))[mz_order]
    new_sp <- x[[main]]
    # ...
}
```

BTW: even in the current code we have to test that `main` is not out of range.

Reply:
OK, now I got it. Still, I would like to keep the `unionPeaks` arguments. With your version users don't have the control to choose the `Spectrum` in `x` from which they want to keep all additional fields except `mz` and `intensity` if they want to get the union of all peaks. For peak union it will always be the middle spectrum. I could however imagine users wanting to choose also the *main* spectrum in that case.

I'll fix the test for `main`, thanks!

Reply:
> I could however imagine users wanting to choose also the main spectrum in that case.

Ok, valid point. So choose `unionPeaks` please. Sorry for the long discussion about this.

Reply:
Discussing is important. I'm glad to get your valuable feedback on each PR.